### PR TITLE
Refactoring installation

### PR DESCRIPTION
**Pull request by jderusse**
Refactor the way dns-gen should be installed.

- no more dnsmasq to install locally
- just start the container and prepend the /etc/resolv.conf file
- provide an install script to ease setup

Tested:
- [x] debian9
- [ ] debian 8
- [ ] ubuntu 16
- [x] ubuntu 18
- [ ] distrib with custom dnsmasq installed
- [x] distrib with previous installation of dns-gen

### CODE REVIEW

**Comment by mykiwi** [comment]
- [ ] archlinux / manjaro

:angel:

File: bin/install [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+  do
+    case $answer in
+      ([yY][eE][sS] | [yY]) return 0;;
+      ([nN][oO] | [nN]) return 1;;
+      (*) echo "${C4}${message}${C3} yes ${C4}or${C3} no${C0}";;
+    esac
+  done
+}
+
+main
```
Comment (flobb) [inline bin/install]: Missing return line

**Review by flobb — COMMENTED** [review]
It's possible to send the generated list of DNS to the logs for debug purposes ? To have it when we do `docker logs` command.

**Comment by flobb** [comment]
On `Ubuntu 18.04.1 LTS` when executing the `bin\install` : `/sbin/ifconfig: not found`.
After some digging :
`The new and recommended alternative for examining a network configuration on Debian Linux is ip command.`

File: bin/install [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+C0='\033[0m'
+
+if [ "$(id -u)" -ne 0 ]; then
+  echo "${C1}This script must be run by root${C0}"
+  echo "${C2} running ${C3}sudo $0 $@${C0}"
+  sudo $0 "$@"
+  exit 0
+fi
+
+GATEWAY=$(/sbin/ifconfig docker0 | grep "inet" | head -n1 | awk '{ print $2}' | cut -d: -f2 || echo "")
```
Comment (flobb) [inline bin/install]: Or `ip -4 addr show docker0 | grep -oP '(?<=inet\s)\d+(\.\d+){3}'` for latest Debian based system (if ifconfig not available)

**Comment by flobb** [comment]
`netplan` instead of `resolvconf` ? :x

File: bin/install (hunk 2 above)
Comment (jderusse) [reply bin/install]: thank you. Fixed

File: bin/install (hunk 1 above)
Comment (jderusse) [reply bin/install]: done

**Comment by flobb** [comment]
ça fonctionne sous ubuntu 18.04 LTS
```
> bin/install
This script must be run by root
 running sudo bin/install 
Requirements
 - resolvconf is installed
 - docker is installed
 - container dns-gen stoped
 - resolv.conf cleaned
 - interface docker0 exists
 - port 53 is free
Install
2: Pulling from jderusse/dns-gen
Digest: sha256:3057041e527f8ed5f78da663c2d5b3c735aaaa87f2bd15b874693379d368e915
Status: Image is up to date for jderusse/dns-gen:2
 - started container dns-gen
 - appended nameserver 172.17.0.1 # dns-gen in /etc/resolvconf/resolv.conf.d/head

> dig dig beintranet_front_1.docker                                                                                                                
; <<>> DiG 9.11.3-1ubuntu1.3-Ubuntu <<>> dig beintranet_front_1.docker
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 55084
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;dig.				IN	A

;; Query time: 0 msec
;; SERVER: 172.17.0.1#53(172.17.0.1)
;; WHEN: Tue Feb 19 11:36:33 CET 2019
;; MSG SIZE  rcvd: 32

;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 20995
;; flags: qr aa rd ra ad; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;beintranet_front_1.docker.	IN	A

;; ANSWER SECTION:
beintranet_front_1.docker. 0	IN	A	172.18.0.3

;; Query time: 0 msec
;; SERVER: 172.17.0.1#53(172.17.0.1)
;; WHEN: Tue Feb 19 11:36:33 CET 2019
;; MSG SIZE  rcvd: 70

```

File: bin/install [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+install_container () {
+  docker pull jderusse/dns-gen:2
+  docker run -d --name dns-gen \
+        --restart always \
+        --log-opt "max-size=10m" \
+        --net host \
+        -e GATEWAY=$GATEWAY \
+        --volume /:/host \
```
Comment (flobb) [inline bin/install]: Why do you mount all the FS ???

File: README.md [hunk 4]
```diff
@@ -4,190 +4,126 @@ dns-gen sets up a container running Dnsmasq and [docker-gen].
...
+**One more thing** When you start your host, the docker service is not fully
+loaded.
+Until this daemon is loaded, the dns container will not be automatically started
+and you will notice bad performance when your host will try to resolve DNS.
+The service is not fully loaded, because it uses a feature of systemd called
+[socket activation]: The first access to the docker socket will trigger the
+start of the true service.
+To skip this feature, you simply have to activate the docker service.
 
+    sudo update-rc.d docker enable
```
Comment (flobb) [inline README.md:142]: ## Uninstall

`docker update --restart=no dns-gen && docker stop dns-gen`

File: bin/install (hunk 3 above)
Comment (jderusse) [reply bin/install]: To be able to read the host's `/etc/resolv.conf` file (I added a Q/A section in readme to get a full complet answer to your question)

File: README.md (hunk 4 above)
Comment (jderusse) [reply README.md:142]: I'm not sure to get your comment. I added a line in the section `Previous version of docker-gen`

Comment (flobb) [reply README.md:142]: c'est pour desinstaller le container si on n'en pas plus besoin, juste un stop n'est pas suffisant

**Pull request merged**